### PR TITLE
Internalize `Process.fork`

### PR DESCRIPTION
**Pull request by straight-shoota**
This is a continuation of https://github.com/crystal-lang/crystal/pull/9136 which changed `Process.fork` to `:nodoc:`. But it still *exists* in the `Process` namespace, even if it doesn't show up in the docs.
`fork` is a system-specific implementation detail and this patch moves it into the `Crystal::System` namespace.

If you *really*, *really* want to use `fork` despite being dangerous and non-portable, you can *technically* still continue to use `Crystal::System::Process.fork`. But then it's at least clearly shows that it's not part of any public API.

Resolves https://github.com/crystal-lang/crystal/issues/6421

Ref https://forum.crystal-lang.org/t/deprecate-top-level-fork/5130/7

### CODE REVIEW

File: src/crystal/system/unix/process.cr [hunk 1]
```diff
@@ -116,6 +116,28 @@ struct Crystal::System::Process
...
+        yield
+        LibC._exit 0
+      rescue ex
+        ex.inspect_with_backtrace STDERR
+        STDERR.flush
+        LibC._exit 1
+      ensure
+        LibC._exit 254 # not reached
+      end
+    end
```
Comment (Sija) [inline src/crystal/system/unix/process.cr:138]: Could be cleaned up a bit:

```suggestion
    {% raise("Process fork is unsupported with multithreaded mode") if flag?(:preview_mt) %}

    if pid = fork
      return pid
    end

    begin
      yield
      LibC._exit 0
    rescue ex
      ex.inspect_with_backtrace STDERR
      STDERR.flush
      LibC._exit 1
    ensure
      LibC._exit 254 # not reached
    end
```

File: src/compiler/crystal/compiler.cr [hunk 2]
```diff
@@ -477,7 +479,7 @@ module Crystal
               end
             end
 
-            codegen_process = Process.fork do
+            codegen_process = Crystal::System::Process.fork do
```
Comment (HertzDevil) [inline src/compiler/crystal/compiler.cr:482]: I think `codegen_process` needs to be wrapped inside a `::Process`

**Review by HertzDevil — APPROVED** [review]

**Comment by stefandd** [comment]
This leads to `Unhandled exception in spawn: Not Implemented: Process.fork (NotImplementedError)` errors on Windows when trying to compile a debug version with multiple threads. It does *not* create this error when trying to compile the release version .

This works
```
crystal build --release --threads 4 matmul.cr
   Creating library c:\Crystal\prj\matmul.lib and object c:\Crystal\prj\matmul.exp
```

This generates a number of the above exceptions:
```
crystal build -d --threads 4 matmul.cr
Unhandled exception in spawn: Not Implemented: Process.fork (NotImplementedError)
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277238 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277134 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277098 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +116687 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +2081928 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +1613161 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +146688 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +1892397 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +130969 in '??'
Unhandled exception in spawn: Not Implemented: Process.fork (NotImplementedError)
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277238 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277134 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277098 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +116687 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +2081928 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +1613161 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +146688 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +1892397 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +130969 in '??'
Unhandled exception in spawn: Not Implemented: Process.fork (NotImplementedError)
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277238 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277134 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277098 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +116687 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +2081928 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +1613161 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +146688 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +1892397 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +130969 in '??'
Unhandled exception in spawn: Not Implemented: Process.fork (NotImplementedError)
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277238 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277134 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277098 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +116687 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +2081928 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +1613161 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +146688 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +1892397 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +130969 in '??'
Unhandled exception in spawn: Not Implemented: Process.fork (NotImplementedError)
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277238 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277134 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +277098 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +116687 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +2081928 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +1613161 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +147648 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +1892397 in '??'
  from C:\Users\mannsfel\scoop\apps\crystal\current\crystal.exe +130969 in '??'
```

**Comment by straight-shoota** [comment]
Are you sure that behaviour has changed with this PR? It shouldn't include anything that would affect multithreading behaviour on windows. And in fact I see crystal 1.7.2 behaves exactly the same as what you are reporting.
Note that this patch is not yet released. It's only effective in a 1.8.0-dev compiler (nightly build or self compiled).
The `--threads` option to the compiler didn't work before this PR and it doesn't after.

**Pull request merged**